content/blog/eng/careful-with-that-constant-developer.md

```diff
@@ -0,0 +1,462 @@
+{
+    public function useTheForce(): string 
+    {
+        return Jedi::SIDE_OF_THE_FORCE;
+    }
+}
+```
+
+This kind of approach is one of the alternatives you can use to avoid [Constant interface](https://en.wikipedia.org/wiki/Constant_interface) anti-pattern.  
+And precisely, it brings us to speak of interfaces.  
+
+# Interface constants
+
+Indeed PHP allows you to use interface constants too.  
+According to [PHP documentation](https://www.php.net/manual/en/language.oop5.interfaces.php#language.oop5.interfaces.constants), it's possible for interfaces to have constants and they can't be overridden by a class/interface that inherits them.  
+So you can define an interface constant, necessarily `public`, like this
+
+```php
+<?php
+declare(strict_types=1);
+
+interface JediInterface 
+{
+    public const SIDE_OF_THE_FORCE = 'light';
+}
+```
+
+But you can't override it, in fact the following code
+
+```php
+<?php
+declare(strict_types=1);
+
+interface JediInterface 
+{
+    public const SIDE_OF_THE_FORCE = 'light';
+}
+
+class Padawan implements JediInterface 
+{
+    public const SIDE_OF_THE_FORCE = 'dark';
+}
+```
+
+outputs a fatal error
+
+```
+Fatal error: Cannot inherit previously-inherited or override constant SIDE_OF_THE_FORCE from interface JediInterface in [...]
+```
+
+So, it seems you came to an end.   
+You have your constant, it can't be changed, it doesn't belong to a particular instance.  
+However, it's not that simple.  
+
+# An unexpected behaviour
+
+Take a look at this code
+
+```php
+<?php
+declare(strict_types=1);
+
+interface JediInterface 
+{
+    public const SIDE_OF_THE_FORCE = 'light';
+}
+
+class Master implements JediInterface 
+{
+}
+
+class Padawan extends Master 
+{
+    public const SIDE_OF_THE_FORCE = 'dark';
+
+    public function useTheForce(): string 
+    {
+        return self::SIDE_OF_THE_FORCE;
+    }
+}
+
+$anakin = new Padawan();
+echo $anakin->useTheForce(); 
+// dark
+```
+
+This piece of code doesn't output any error.  
+It turns out that a derived class from a super class that implements an interface, can override the interface constants, despite the super class can't.  
+And in fact, if you make the derived class implement the interface likewise
+
+```php
+<?php
+declare(strict_types=1);
+
+interface JediInterface 
+{
+    public const SIDE_OF_THE_FORCE = 'light';
+}
+
+class Master implements JediInterface 
+{
+}
+
+class Padawan extends Master implements JediInterface 
+{
+    public const SIDE_OF_THE_FORCE = 'dark';
+}
+```
+
+you get again a fatal error
+
+```
+Fatal error: Cannot inherit previously-inherited or override constant SIDE_OF_THE_FORCE from interface JediInterface in [...]
+```
+
+So, bug or feature?  
+It's a fair question, bearing in mind that the PHPStorm static analysis tool, currently (version 2020.3.2) reports always as an error the attempt to redefine an interface constant, even if it's redefined by a child class that doesn't implement directly the interface.  
```

Comment:
It seems the readers already know (so they can remember) that PHPStorm static analysis tool does something, but we don't know if the know. 
What about this?

This is a fair question, since the PHPStorm static analysis tool, currently (version 2020.3.2), always reports an attempt to redefine an interface constant as an error.

Reply:
It's ok to me!